python/tank_vendor/shotgun_authentication/shotgun_authenticator.py

```diff
@@ -0,0 +1,239 @@
+        """
+        return self._defaults_manager.get_host()
+
+    def get_default_http_proxy(self):
+        """
+        Returns the HTTP proxy from the defaults manager.
+        :returns: The default http proxy string.
+        """
+        return self._defaults_manager.get_http_proxy()
+
+    def get_default_user(self):
+        """
+        Returns the default user from the defaults manager.
+        :returns: A ShotgunUser derived instance if available, None otherwise.
+        """
+        # Get the credentials
+        credentials = self._defaults_manager.get_user()
+
+        # There is no default user.
+        if not credentials:
+            return None
+
+        # If this looks like an api user, delegate to create_script_user.
+        # If some of the arguments are missing, don't worry, create_script_user
+        # will take care of it.
+        if "api_script" in credentials or "api_key" in credentials:
+            return self.create_script_user(
+                api_script=credentials.get("api_script"),
+                api_key=credentials.get("api_key"),
+                host=credentials.get("host"),
+                http_proxy=credentials.get("http_proxy")
+            )
+        # If this looks like a session user, delegate to create_session_user.
+        # If some of the arguments are missing, don't worry, create_session_user
+        # will take care of it.
+        elif "login" in credentials or "password" in credentials or "session_token" in credentials:
+            return self.create_session_user(
+                login=credentials.get("login"),
+                password=credentials.get("password"),
+                session_token=credentials.get("session_token"),
+                host=credentials.get("host"),
+                http_proxy=credentials.get("http_proxy")
+            )
+        # We don't know what this is, abort!
+        else:
+            raise IncompleteCredentials(
+                "unknown credentials configuration: %s" % credentials
+            )
+
+    def get_user(self):
+        """
+        This method will always return a valid user. It will first ask for the
+        default user to the defaults manager. If no user is returned, then a
+        saved user will be retrieved for the default host. If none is found, the
+        you will be prompted to enter login information interactively.
+
+        :returns: A ShotgunUser derived instance.
+        """
+        # Get the default user first for backward compatibility reasons. Toolkit
+        # provides it's own defaults manager which has a get_user that returns
+        # the credentials for the script user in shotgun.yml, so that has to
+        # have precedence over the saved user.
+        user = self.get_default_user() or self.get_saved_user()
+        if user:
+            return user
+
+        # Prompt the client for user credentials and connection information
+        user = self.get_user_from_prompt()
+        user.save()
```

Comment:
Should this also be an internal thing that the user object handles behind the scenes?